src/ui/colorui.js

```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
 				editor.editing.view.focus();
 			} );
 
+			if ( documentColorsCount !== 0 ) {
+				dropdownView.on( 'change:isOpen', ( evt, name, val ) => {
+					if ( val ) {
+						this.findAndSetDocumentColors();
+					}
+				} );
+			}
+
 			return dropdownView;
 		} );
 	}
+
+	/**
+	 * Method scans through editor's content and search for text node attributes with name defined in {@link #commandName}.
+	 * Found entries are set as document colors.
+	 *
+	 * All previously stored document colors will be lost with that process.
+	 *
+	 * Method is used to determines document colors when UI dropdown is opened.
+	 *
+	 * @private
+	 */
+	findAndSetDocumentColors() {
+		const model = this.editor.model;
+		const document = model.document;
+		const maxCount = this.colorTableView.documentColorsCount;
+		const documentColors = this.colorTableView.documentColors;
+
+		documentColors.clear();
+
+		for ( const rootName of document.getRootNames() ) {
+			const root = document.getRoot( rootName );
+			const range = model.createRangeIn( root );
+			for ( const node of range.getItems() ) {
+				if ( node.is( 'textProxy' ) && node.hasAttribute( this.componentName ) ) {
+					this.addColorToDocumentColors( node.getAttribute( this.componentName ) );
+					if ( documentColors.length >= maxCount ) {
+						return;
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Method tries to find a color on predefined color list to prevent of creating custom color from it.
```

Comment:
It's enough to say:

```suggestion
	 * Method adds the `color` to document colors list. If possible, it will attempt to use data from the {@link #colorDefinitions} (label, color options).
```

The rest of header description should be removed.

Your proposal included too much details and a little bit of repetition.